prusti-specs/src/specifications/preparser.rs

```diff
@@ -289,557 +58,547 @@ impl Parse for SpecEntArgs {
-            self.expr.push(TokenTree::Group(group));
+        } else {
+            false
         }
-        Ok(())
     }
-    fn resolve_rust_expr(&mut self) -> syn::Result<()> {
-        // if nothing of the above happened, we just continue parsing as a Rust expression
-        if self.expected_only_operator {
-            self.input.pop();
-            return Err(self.error_expected_operator());
+    /// does the given operator appear in the stream at top level
+    fn contains_operator(&self, stream: &VecDeque<TokenTree>, operator: &str) -> bool {
+        let mut stream = stream.clone();
+        while !stream.is_empty() {
+            if self.peek_operator_stream(&stream, operator) {
+                return true;
+            }
+            stream.pop_front();
         }
-
-        let token = self.input.pop().unwrap();
-        self.expr.push(token);
-        self.expected_operator = true;
-        Ok(())
+        false
     }
-    /// Creates a single Prusti assertion from the input and returns it.
-    pub fn extract_assertion(&mut self) -> syn::Result<AssertionWithoutId> {
-        // detect possibly ambiguous input
-        if self.input.contains_both_and_or() {
-            return Err(self.error_ambiguous_expression());
-        }
-
-        // preparse the input into atomic Prusti assertions
-        while !self.input.is_empty() {
-            if self.input.check_and_consume_operator("&&") {
-                if let Err(err) = self.resolve_and() {
-                    return Err(err);
-                }
-            }
-            else if self.input.check_and_consume_operator("==>") {
-                return self.resolve_implies();
+    /// does the given operator appear in the stream anywhere
+    fn contains_operator_recursive(&self, stream: &VecDeque<TokenTree>, operator: &str) -> Option<Span> {
+        let mut stream = stream.clone();
+        while !stream.is_empty() {
+            if self.peek_operator_stream(&stream, operator) {
+                return Some(self.operator_span(&stream, operator));
             }
-            else if self.input.check_and_consume_keyword("forall") {
-                if let Err(err) = self.resolve_forall() {
-                    return Err(err);
+            if let Some(TokenTree::Group(group)) = stream.front() {
+                let nested_stream: VecDeque<TokenTree> = group.stream().into_iter().collect();
+                if let Some(span) = self.contains_operator_recursive(&nested_stream, operator) {
+                    return Some(span);
                 }
             }
-            else if self.input.check_and_consume_operator("|=") {
-                if let Err(err) = self.resolve_spec_ent() {
-                    return Err(err);
-                }
+            stream.pop_front();
+        }
+        None
+    }
+    /// get the span of the peeked operator
+    fn operator_span(&self, stream: &VecDeque<TokenTree>, operator: &str) -> Span {
+        let mut span: Option<Span> = None;
```

Comment:
```suggestion
        (0..operator.len())
            .filter_map(|i| stream.get(i))
            .map(|character| character.span())
            .reduce(|span_a, span_b| span_a.join(span_b).unwrap())
```
I also wonder if it is not enough to `join` the span of the first and last characters.

Reply:
I think that's why span joins produce `Option`s - if you join two non-adjacent spans, you get `None`.

Reply:
Nevermind - you can join two non-adjacent spans.

Reply:
Addressed by c77b6d0b1638c36b81eddb1e87af37c3f1fff212